### PR TITLE
Add support for recording ethtool stats

### PR DESCRIPTION
**Pull request by mmynk**
The PR is broken into modular commits, so it might be easier to review the PR commit-by-commit.

Commit: [Add crate for reading NIC stats using ethtool](https://github.com/facebookincubator/below/commit/965795c0c1b2675cc22cf781a80f5e06cdbea50a)
  - Adds the library support to read and parse `ethtool` stats using `ioctl`
  - Parses some common stats into fields while others are stored as raw stats

Commit: [Add model to represent ethtool stats](https://github.com/facebookincubator/below/commit/66dbb26ab612d3f87131aabbbf6bd33112234c90)
  - Adds the model struct and parsing logic to convert `ethtool` stats into a sub-model in existing `network.interface` model

Commit: [Add render configs for ethtool fields](https://github.com/facebookincubator/below/commit/9401fed6922f1c843d1ae0ed3e1949cf37b26fad)
  - Updates and adds the required commands for dumping new fields and model
  - Adds the parsing for rendering the new fields and model

Commit: [Add ability for printing raw_stats in OpenMetrics format](https://github.com/facebookincubator/below/commit/7d6349efa2d1d72795d3e6a18417384e44113fb2)
  - Modifies the printing logic of `OpenMetrics` format to handle a map type

### CODE REVIEW

**Comment by mmynk** [comment]
@dschatzberg no worries, take your time. thanks for the ack!

File: below/ethtool/src/ethtool.rs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    Ok(values)
+}
+
+pub trait EthtoolReadable {
+    fn new(if_name: &str) -> Result<Self, Error> where Self: Sized;
+    fn stats(&self) -> Result<Vec<(String, u64)>, Error>;
+}
+
+pub struct Ethtool {
+    sock_fd: i32,
```
Comment (dschatzberg) [inline below/ethtool/src/ethtool.rs]: You should use https://doc.rust-lang.org/stable/std/os/fd/struct.OwnedFd.html so you don't need your own Drop impl

File: below/ethtool/src/test.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            ("queue_1_tx_tx_poll", 45366),
+            ("queue_1_tx_bad_req_id", 0),
+            ("queue_1_tx_llq_buffer_copy", 0),
+            ("queue_1_tx_missed_tx", 0),
+            ("queue_1_tx_unmask_interrupt", 45365),
+            ("queue_1_rx_cnt", 25191),
+            ("queue_1_rx_bytes", 6562216),
+            ("queue_1_rx_refil_partial", 0),
+            ("queue_1_rx_csum_bad", 0),
+            ("queue_1_rx_page_alloc_fail", 0),
```
Comment (dschatzberg) [inline below/ethtool/src/test.rs:47]: Where is the format of these keys specified? I tested on a couple machine and I don't see this `queue_#_##` format at all.

**Review by dschatzberg — CHANGES_REQUESTED** [review]

**Review by dschatzberg — CHANGES_REQUESTED** [review]
A lot of this commit should be squashed into the one before, the model changes look reasonable

File: below/ethtool/src/ethtool.rs [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+                    offset
+                )))
+            }
+        }
+    }
+
+    Ok(values)
+}
+
+pub trait EthtoolReadable {
```
[the remark this answers is not in the document]
Comment (mmynk) [reply below/ethtool/src/ethtool.rs]: yes this trait is introduced for easier testing

File: below/ethtool/src/ethtool.rs [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    };
+    let mut request = IfReq {
+        if_name: ifname,
+        if_data: data,
+    };
+
+    let exit_code = unsafe { ioctl(fd, nix::libc::SIOCETHTOOL, &mut request) };
+
+    if exit_code != 0 {
+        println!("code: {exit_code}, data: {:?}", request.if_data);
```
[the remark this answers is not in the document]
Comment (mmynk) [reply below/ethtool/src/ethtool.rs]: oh yes, it was a debug print
thanks for catching!

File: below/ethtool/src/test.rs (hunk 2 above)
Comment (mmynk) [reply below/ethtool/src/test.rs:47]: ubuntu 18/22 on AWS has the format `queue_#_*`
for v1, i thought of supporting only this format as I couldn't find an exhaustive list of formats readily. in the next PR, we will add support for more formats and improve the translation logic for parsing.
regardless, stats which are not directly parseable are stored as `raw_stats`.

File: below/ethtool/src/ethtool.rs [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    pub len: u32,
+    pub data: [u8; MAX_GSTRINGS * ETH_GSTATS_LEN],
+}
+
+#[derive(Debug)]
+#[repr(C)]
+struct IfReq {
+    if_name: [u8; IFNAME_MAX_SIZE],
+    if_data: usize,
+}
```
[the remark this answers is not in the document]
Comment (mmynk) [reply below/ethtool/src/ethtool.rs]: do you suggest committing that header and integrating `bindgen` into the build or committing the generated rust bindings?

File: below/ethtool/src/ethtool.rs [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+            Err(errno) => Err(Error::GStringsReadError(errno)),
+        }
+    }
+
+    /// Get the statistics for the features using EHTOOL_GSTATS command
+    fn gstats(&self, features: &Vec<String>) -> Result<Vec<u64>, Error> {
+        let length = features.len();
+        let mut gstats = GStats {
+            cmd: ETHTOOL_GSTATS,
+            len: features.len() as u32,
```
[the remark this answers is not in the document]
Comment (mmynk) [reply below/ethtool/src/ethtool.rs]: `features` is the out-parameter of the `self.gstrings()` call and `self.gstats()` uses the corresponding `features.len()`.

**Comment by mmynk** [comment]
should I squash commits in the PR or is it okay to do so while merging?

File: below/ethtool/src/ethtool.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dschatzberg) [reply below/ethtool/src/ethtool.rs]: > do you suggest committing that header and integrating bindgen into the build

Let's do this one.

**Comment by dschatzberg** [comment]
Sorry, Github doesn't make it clear that I was referring to the 2nd commit. Please squash in the PR - to be clear, I'm fine with multiple commits per PR, in fact, I prefer it. But, many of the changes in the 2nd commit are logically more a part of the first commit which makes it harder to review.

Personally, I like to rewrite my commits after writing the full PR - e.g. via https://git-scm.com/book/en/v2/Git-Tools-Interactive-Staging

**Comment by mmynk** [comment]
@dschatzberg Thanks for the detailed review, have addressed most of the comments. Could you please take a look again?
Build is failing because of recent `dateutil` changes, fixed in #8206

File: below/ethtool/src/ethtool.rs [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+const ETHTOOL_GSTATS: u32 = 0x1d;
+const ETH_SS_STATS: u32 = 0x1;
+const ETH_GSTRING_LEN: usize = 32;
+const ETH_GSTATS_LEN: usize = 8;
+
+/// Maximum size of an interface name
+const IFNAME_MAX_SIZE: usize = 16;
+
+/// MAX_GSTRINGS maximum number of stats entries that ethtool can retrieve
+const MAX_GSTRINGS: usize = 8192;
```
[the remark this answers is not in the document]
Comment (mmynk) [reply below/ethtool/src/ethtool.rs]: yes, and it is in fact dynamic
this is a just an upper limit for creating the initial buffer

File: below/ethtool/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    pub fn read_stats<T: reader::EthtoolReadable>(&self) -> Result<EthtoolStats> {
+        let mut nic_stats_map = BTreeMap::new();
+        let if_names = match self.read_interfaces() {
+            Ok(ifs) => ifs,
+            Err(err) => {
+                return Err(err);
+            }
+        };
```
Comment (dschatzberg) [inline below/ethtool/src/lib.rs:124]: Do you need this map clause? Can't `?` work to pass along the error?

File: below/ethtool/src/reader.rs [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+const MAX_GSTRINGS: usize = 8192;
+
+#[derive(Debug)]
+#[repr(C)]
+struct StringSetInfo {
+    cmd: u32,
+    reserved: u32,
+    mask: u32,
+    data: usize,
+}
```
Comment (dschatzberg) [inline below/ethtool/src/reader.rs:33]: Isn't this defined by ethtool.h? Why redefine it?

File: below/ethtool/src/reader.rs [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+        }
+    }
+
+    /// Get the feature names using ETHTOOL_GSTRINGS command
+    fn gstrings(&self, length: usize) -> Result<Vec<String>, Error> {
+        let mut gstrings = GStrings {
+            cmd: ETHTOOL_GSTRINGS,
+            string_set: ETH_SS_STATS,
+            len: length as u32,
+            data: [0u8; MAX_GSTRINGS * ETH_GSTRINGS_LEN],
```
Comment (dschatzberg) [inline below/ethtool/src/reader.rs:170]: Why are you allocating MAX_GSTRINGS * ETH_GSTRINGS_LEN and not len * ETH_GSTRINGS_LEN?

File: below/ethtool/src/reader.rs [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+    pub len: u32,
+    pub data: [u8; MAX_GSTRINGS * ETH_GSTATS_LEN],
+}
+
+#[derive(Debug)]
+#[repr(C)]
+struct IfReq {
+    if_name: [u8; IFNAME_MAX_SIZE],
+    if_data: usize,
+}
```
Comment (dschatzberg) [inline below/ethtool/src/reader.rs:57]: All of these should come from the corresponding C header, no?

File: below/ethtool/src/reader.rs (hunk 10 above)
Comment (dschatzberg) [reply below/ethtool/src/reader.rs:170]: https://github.com/torvalds/linux/blob/master/include/uapi/linux/ethtool.h#L829

The ethtool header describes len as an out parameter with the number of strings in the set so I'm not clear why you're setting it here.

File: below/ethtool/src/reader.rs [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+    }
+
+    /// Get the statistics for the features using ETHTOOL_GSTATS command
+    fn gstats(&self, features: &[String]) -> Result<Vec<u64>, Error> {
+        let length = features.len();
+        let mut gstats = GStats {
+            cmd: ETHTOOL_GSTATS,
+            len: features.len() as u32,
+            data: [0u8; MAX_GSTRINGS * ETH_GSTATS_LEN],
+        };
```
Comment (dschatzberg) [inline below/ethtool/src/reader.rs:190]: I have similar concerns here as with gstrings, this doesn't match my understanding of the API

**Review by dschatzberg — CHANGES_REQUESTED** [review]

File: below/ethtool/src/reader.rs (hunk 9 above)
Comment (mmynk) [reply below/ethtool/src/reader.rs:33]: i tried, but i was having trouble figuring out how to work with the struct generated by bindgen
```rs
#[repr(C)]
#[derive(Debug)]
pub struct ethtool_sset_info {
    pub cmd: __u32,
    pub reserved: __u32,
    pub sset_mask: __u64,
    pub data: __IncompleteArrayField<__u32>,
}
```

not sure how to init `__IncompleteArrayField`, will try figuring it out.

File: below/ethtool/src/reader.rs (hunk 10 above)
Comment (mmynk) [reply below/ethtool/src/reader.rs:170]: that's right, however we cannot have a dynamically size array, so i've set it to max length possible
while passing the array, now i'm only passing the slice which would have value, ie: `len * ETH_GSTRINGS_LEN`

File: below/ethtool/src/reader.rs (hunk 11 above)
Comment (mmynk) [reply below/ethtool/src/reader.rs:29]: I don't think a corresponding struct is available in `ethtool.h`

**Comment by mmynk** [comment]
@dschatzberg have removed all the re-defined structs and used structs from the binding, this led to a few `unsafe` calls, hope that's okay?[1]

[1] https://github.com/facebookincubator/below/pull/8204/commits/c4cd7ea0fcfecab771c2523c69a7788d6587a12c

**Comment by mmynk** [comment]
@dschatzberg could you please re-review the PR? thanks!

**Comment by dschatzberg** [comment]
@mmynk Sorry I've been behind on some other work. I'm out on vacation all next week so unlikely I'll get to look at this for the next week but I'll review once I'm back if not.

**Comment by mmynk** [comment]
hey @dschatzberg, just wanted to drop a gentle reminder. thanks!

**Comment by dschatzberg** [comment]
@mmynk Sorry for the delay, I'm working on the review and should get you feedback this week

File: below/ethtool/src/reader.rs [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+
+        let n_stats = length as u32;
+
+        // Allocate memory for the struct
+        let stats_ptr = unsafe { alloc::alloc(layout) } as *mut ethtool_stats;
+
+        // Initialize the fields of the struct
+        unsafe {
+            let cmd_ptr = ptr::addr_of_mut!((*stats_ptr).cmd);
+            let n_stats_ptr = ptr::addr_of_mut!((*stats_ptr).n_stats);
```
Comment (dschatzberg) [inline below/ethtool/src/reader.rs:224]: Same thing here - this is an output parameter

**Review by dschatzberg — CHANGES_REQUESTED** [review]
This is looking great - I have a couple inline questions regarding what look like output parameters to me, but otherwise this is just about ready to be merged.

**Comment by mmynk** [comment]
@dschatzberg thank you! have addressed the comments.

**Comment by mmynk** [comment]
@dschatzberg gentle reminder, thanks!

**Comment by dschatzberg** [comment]
@mmynk Just a heads up, I changed a couple things in the PR when integrating it internally:

1) There's a config option now to enable gathering ethtool stats - we need this for rollout safety so we can push code and dynamically turn things off/on separately. It defaults to disable but if we have no issues rolling it out I'd be happy to flip that.

2) I ended up committing the bindgen'd file itself because there's no pre-existing library that includes the ethtool header so it made it a bit challenging to generate in our repository where all dependencies must be vendored.

**Comment by mmynk** [comment]
@dschatzberg Do you see value in syncing those changes in upstream?

**Comment by dschatzberg** [comment]
@mmynk I already did sync those changes (e.g. https://github.com/facebookincubator/below/commit/7fc6f7f7e1bdee0716a49bf2ab4a44e1d3623dc7#diff-080793436ea7d5d338322a440a1430f13a281c6e832366a2f03ea641a44c797eR63) as part of commiting this PR.

Just so it's clear - the source of truth for Below is Meta's internal code repository and we have automation to sync changes from that to this repository on Github. Every PR you submit has to land first in our internal code repository before it makes it to Github. So the term "upstream" can be a bit confusing.

**Pull request closed without merging**